src/background.js

```diff
@@ -253,10 +262,43 @@ async function onResultsRequested(query) {
  * browser.urlbar.onResultPicked listener.  Called when a tip button is picked.
  */
 async function onResultPicked(payload) {
+  let tip = payload.type;
+
+  // Set tipPicked so our onEngagement listener knows a tip was picked.
+  tipPicked = true;
+
   // Update picked-count telemetry.
-  browser.telemetry.keyedScalarAdd(TELEMETRY_PICKED, payload.type, 1);
+  browser.telemetry.keyedScalarAdd(TELEMETRY_PICKED, tip, 1);
+
+  // We open a survey 100% of the time a tip is picked.  If the browser will not
+  // restart due to the user's picking the tip, we open the survey now (below).
+  // If the browser will restart, we open the survey after restart (in enroll).
 
-  switch (payload.type) {
+  // Determine whether the browser will restart.  For REFRESH and UPDATE_REFRESH
+  // we show the profile-reset dialog, which the user can cancel, but for our
+  // purposes we assume the browser will restart.
+  let willRestart = false;
+  switch (tip) {
+    case TIPS.REFRESH:
+    case TIPS.UPDATE_ASK:
+    case TIPS.UPDATE_REFRESH:
+    case TIPS.UPDATE_RESTART:
+      willRestart = true;
+      break;
+  }
+
+  // If we're restarting, save the picked tip type in storage.
+  if (willRestart) {
+    browser.storage.local.set({ surveyPickedTip: tip });
+  }
+
+  // Do the tip action.
+  //
+  // IMPORTANT: Don't `await` anything before this!  Some of these functions are
```

Comment:
must this function be declared async? I know the API is async, but maybe you could declare this non async and move this comment to the function declaration explaining this can't be async. await from the outside works regardless.

Reply:
Good idea